src/protocol/boolean/prefix_or.rs

```diff
@@ -0,0 +1,400 @@
+
+        Ok(b[0..self.input.len()].to_vec())
+    }
+
+    /// This method takes a slice of bits of the length `l`, add `m` dummy
+    /// bits to the end of the slice, and returns it as a new vector. The
+    /// output vector's length is `λ^2` where `λ = sqrt(l + m) ∈ Z`.
+    fn add_dummy_bits(a: &[Replicated<B>]) -> (Vec<Replicated<B>>, usize) {
+        // We plan to use u32, which we'll add 4 dummy bits to get λ = 6.
+        // Since we don't want to compute sqrt() each time this protocol
+        // is called, we'll assume that the input is 32-bit long.
+        // We can modify this function if we need to support other lengths.
+        let l = a.len();
+        let lambda: usize = match l {
+            8 => 3,
+            16 => 4,
+            32 => 6,
+            _ => panic!("bit length must 8, 16 or 32"),
+        };
+        let dummy = vec![Replicated::new(B::ZERO, B::ZERO); lambda * lambda - l];
+        ([a, &dummy].concat(), lambda)
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+enum Step {
+    BitwiseOrPerBlock,
+    BlockWisePrefixOr,
+    InnerProduct,
+    GetFirstBlockWithOne,
+    SetFirstBlockWithOne,
+}
+
+impl crate::protocol::Step for Step {}
+
+impl AsRef<str> for Step {
+    fn as_ref(&self) -> &str {
+        match self {
+            Self::BitwiseOrPerBlock => "bitwise_or_per_block",
+            Self::BlockWisePrefixOr => "block_wise_prefix_or",
+            Self::InnerProduct => "inner_product",
+            Self::GetFirstBlockWithOne => "get_first_block_with_one",
+            Self::SetFirstBlockWithOne => "set_first_block_with_one",
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use futures::future::try_join_all;
+    use rand::{rngs::mock::StepRng, Rng};
+
+    use crate::{
+        ff::{Field, Fp2},
+        protocol::{QueryId, RecordId},
+        secret_sharing::Replicated,
+        test_fixture::{make_contexts, make_world, share, validate_and_reconstruct, TestWorld},
+    };
+
+    use super::PrefixOr;
+
+    #[tokio::test]
+    pub async fn prefix_or() {
+        const BITS: usize = 32;
+        const TEST_TRIES: usize = 100;
+        let world: TestWorld = make_world(QueryId);
+        let ctx = make_contexts::<Fp2>(&world);
+        let mut rand = StepRng::new(1, 1);
+        let mut rng = rand::thread_rng();
+
+        // Test 32-bit bitwise shares with randomly distributed bits, for 100 times.
+        // The probability of i'th bit being 0 is 1/2^i, so this test covers inputs
+        // that have all 0's in 6-7 first bits.
+        for i in 0..TEST_TRIES {
+            let len = BITS;
+            let input: Vec<Fp2> = (0..len).map(|_| Fp2::from(rng.gen::<bool>())).collect();
+            let mut expected: Vec<Fp2> = Vec::with_capacity(len);
+
+            // Calculate Prefix-Or of the secret number
+            input.iter().fold(Fp2::ZERO, |acc, &x| {
+                expected.push(acc | x);
+                acc | x
+            });
```

Comment:
Is there a way to generate this in one pass, instead of mutating the vector in each iteration? I think there is some `fold` like method for that.

Reply:
Yeah... I can't figure it out either. Whatever, this is just test code =).